### PR TITLE
feat: constant folding for list operations

### LINKED ISSUE #774: Constant folding for collections extensions

### PR DESCRIPTION
**Pull request by ss2165**
+ utility enums and structs for dealing with list ops and types

Closes #774

### CODE REVIEW

File: src/std_extensions/collections.rs [hunk 1]
```diff
@@ -126,22 +193,107 @@ fn list_and_elem_type(list_type_def: &TypeDef) -> (Type, Type) {
     );
     (list_type, elem_type)
 }
+
+/// A list operation
```
Comment (cqc-alec) [inline src/std_extensions/collections.rs:197]: The `collections` extension doesn't seem to be mentioned in the main spec; is there a spec for it somewhere else?

File: src/std_extensions/collections.rs [hunk 2]
```diff
@@ -126,22 +193,107 @@ fn list_and_elem_type(list_type_def: &TypeDef) -> (Type, Type) {
...
+#[derive(Debug, Clone, PartialEq)]
+pub enum ListOp {
+    /// Pop from end of list
```
Comment (cqc-alec) [inline src/std_extensions/collections.rs:200]: What is the behaviour of the program if the list is empty? (Panic? Error? Doesn't seem to be captured in the signature.)

File: src/std_extensions/collections.rs [hunk 3]
```diff
@@ -126,22 +193,107 @@ fn list_and_elem_type(list_type_def: &TypeDef) -> (Type, Type) {
...
+    fn type_args(&self) -> Vec<TypeArg> {
+        vec![TypeArg::Type {
+            ty: self.elem_type.clone(),
+        }]
+    }
+}
+
+impl ListOpInst {
+    /// Convert this list operation to an [`ExtensionOp`] by providing a
+    /// registry to validate the element tyoe against.
```
Comment (cqc-alec) [inline src/std_extensions/collections.rs:262]:
```suggestion
    /// registry to validate the element type against.
```

**Review by cqc-alec — APPROVED** [review]
Mainly question about specification and semantics; the code itself looks fine to me.

File: src/std_extensions/collections.rs (hunk 1 above)
Comment (ss2165) [reply src/std_extensions/collections.rs:197]: Not yet, it is somewhat of a placeholder for adding more list operations. See #725

File: src/std_extensions/collections.rs (hunk 2 above)
Comment (ss2165) [reply src/std_extensions/collections.rs:200]: panics, should be clear in spec once we have it

**Pull request merged**